crates/matrix-sdk-ui/src/timeline/mod.rs

```diff
@@ -369,6 +371,107 @@ impl Timeline {
+        replied_to_info: RepliedToInfo,
+        is_reply: ReplyWithinThread,
+    ) -> Result<(), RoomSendQueueError> {
+        let event_id = replied_to_info.event_id;
+
+        // [The specification](https://spec.matrix.org/v1.10/client-server-api/#user-and-room-mentions) says:
+        //
+        // > Users should not add their own Matrix ID to the `m.mentions` property as
+        // > outgoing messages cannot self-notify.
+        //
+        // If the replied to event has been written by the current user, let's toggle to
+        // `AddMentions::No`.
+        let mention_the_sender = if self.room().own_user_id() == replied_to_info.sender {
+            AddMentions::No
+        } else {
+            AddMentions::Yes
+        };
+
+        let content = match replied_to_info.content {
+            ReplyContent::Message(msg) => {
+                let event = OriginalRoomMessageEvent {
+                    event_id: event_id.to_owned(),
+                    sender: replied_to_info.sender,
+                    origin_server_ts: replied_to_info.timestamp,
+                    room_id: self.room().room_id().to_owned(),
+                    content: msg.to_content(),
+                    unsigned: Default::default(),
+                };
+                content.make_for_thread(&event, is_reply, mention_the_sender)
+            }
+            ReplyContent::Raw(raw_event) => {
+                // Some of the code below technically belongs into ruma. However,
+                // reply fallbacks have been removed in Matrix 1.13 which means
+                // both match arms can use the successor of make_for_thread in
+                // the next ruma release.
+                #[derive(Deserialize)]
+                struct ContentDeHelper {
+                    #[serde(rename = "m.relates_to")]
+                    relates_to: Option<ruma::events::room::encrypted::Relation>,
+                }
+
+                let previous_content =
+                    raw_event.get_field::<ContentDeHelper>("content").ok().flatten();
+                let relates_to = previous_content.as_ref().and_then(|c| c.relates_to.as_ref());
+
+                let mut content = if is_reply == ReplyWithinThread::Yes {
+                    content.make_reply_to_raw(
+                        &raw_event,
+                        event_id.to_owned(),
+                        self.room().room_id(),
+                        ForwardThread::No,
+                        mention_the_sender,
+                    )
+                } else {
+                    content.into()
+                };
+
+                let thread_root =
+                    if let Some(ruma::events::room::encrypted::Relation::Thread(Thread {
+                        event_id,
+                        ..
+                    })) = &relates_to
+                    {
+                        event_id.to_owned()
+                    } else {
+                        event_id.to_owned()
+                    };
```

Comment:
ogod this is the most evil use of shadowing i've seen in a while :D let's get rid of the top-level `event_id` local variable, and use `replied_to_info.event_id` in the else arm :)

Reply:
Guilty. 😅

Reply:
Also destructuring the `Thread` inside the `if let` seems harder to read than not doing it.

Reply:
No, I mean instead of

```rust
if let Some(encrypted::Relation::Thread(Thread { event_id, .. })) = &relates_to {
    event_id.to_owned()
}
```

you could write

```rust
if let Some(encrypted::Relation::Thread(thread)) = &relates_to {
    thread.event_id.to_owned()
}
```

which would be easier to read.

Reply:
Oh. Yes, that's better indeed. 👍